### PR TITLE
proxyd: Add sender-based rate limiter

### PR DESCRIPTION
**Pull request by mslipper**
This PR adds a sender-based rate limiter to proxyd. The new rate limiter throttles eth_sendRawTransaction requests based on sender in order to prevent abusive behavior on the sequencer.

Integration tests have been added to test this functionality.

It is safe to deploy this without changes to the proxyd config - the new limiter will just be disabled. To enable the limiter, add the following stanza to `proxyd.toml`:

```toml
[sender_rate_limit]
enabled = true
interval = "1s"
limit = 1
```

### CODE REVIEW

File: proxyd/server.go [hunk 1]
```diff
@@ -575,6 +596,54 @@ func (s *Server) isUnlimitedUserAgent(origin string) bool {
...
+	// Inflates a types.Transaction object from the transaction's raw bytes.
+	tx := new(types.Transaction)
+	if err := tx.UnmarshalBinary(data); err != nil {
+		log.Debug("could not unmarshal transaction", "err", err, "req_id", GetReqID(ctx))
+		return ErrInvalidParams(err.Error())
+	}
+
+	// Convert the transaction into a Message object so that we can get the
+	// sender. This method performs an ecrecover, which can be expensive.
+	msg, err := tx.AsMessage(types.NewLondonSigner(tx.ChainId()), nil)
```
Comment (Inphi) [inline proxyd/server.go:628]: nit: Can use types.LatestSignerForChainId instead for forwards-compatibility with future tx types

**Review by Inphi — APPROVED** [review]

**Review by zhwrd — APPROVED** [review]

File: proxyd/config.go [hunk 2]
```diff
@@ -104,21 +104,28 @@ type BatchConfig struct {
 	ErrorMessage string `toml:"error_message"`
 }
 
+type SenderRateLimitConfig struct {
```
Comment (tynes) [inline proxyd/config.go:107]: nit: godoc on this struct so its obvious what the config does

File: proxyd/server.go [hunk 3]
```diff
@@ -409,6 +419,17 @@ func (s *Server) handleBatchRPC(ctx context.Context, reqs []json.RawMessage, isL
 			continue
 		}
 
+		// Apply a sender-based rate limit if it is enabled. Note that sender-based rate
+		// limits apply regardless of origin or user-agent. As such, they don't use the
+		// isLimited method.
+		if parsedReq.Method == "eth_sendRawTransaction" && s.senderLim != nil {
```
Comment (tynes) [inline proxyd/server.go:425]: Did you mean to add this only to the batch rpc path?

File: proxyd/server.go [hunk 4]
```diff
@@ -135,6 +140,10 @@ func NewServer(
 			return nil, err
 		}
 	}
+	var senderLim FrontendRateLimiter
+	if senderRateLimitConfig.Enabled {
+		senderLim = limiterFactory(time.Duration(senderRateLimitConfig.Interval), senderRateLimitConfig.Limit, "senders")
```
Comment (tynes) [inline proxyd/server.go:145]: We should probably disallow config that will brick accepting transactions, so interval or limit cannot be 0

File: proxyd/server.go (hunk 3 above)
Comment (mslipper) [reply proxyd/server.go:425]: Proxyd makes everything into a batch so this was intended, yeah.

**Pull request merged**